numpy/lib/tests/test_io.py

```diff
@@ -2343,6 +2343,36 @@ def test_auto_dtype_largeint(self):
         assert_equal(test['f1'], 17179869184)
         assert_equal(test['f2'], 1024)
 
+    def test_unpack_structured(self):
+        # Regression test for gh-4341
+        # Unpacking should work on structured arrays
+        txt = TextIO("M 21 72\nF 35 58")
+        dt = {'names': ('a', 'b', 'c'), 'formats': ('|S1', '<i4', '<f4')}
+        a, b, c = np.genfromtxt(txt, dtype=dt, unpack=True)
+        assert_(a.dtype.str == '|S1')
+        assert_(b.dtype.str == '<i4')
+        assert_(c.dtype.str == '<f4')
+        assert_array_equal(a, np.array([b'M', b'F']))
+        assert_array_equal(b, np.array([21, 35]))
+        assert_array_equal(c, np.array([72.,  58.]))
+
+    def test_unpack_auto_dtype(self):
+        # Regression test for gh-4341
+        # Unpacking should work when dtype=None
+        txt = TextIO("21 72\n35 58")
+        a, b = np.genfromtxt(txt, dtype=None, unpack=True)
+        assert_array_equal(a, np.array([21, 35]))
+        assert_array_equal(b, np.array([72, 58]))
+
+    def test_squeeze_scalar(self):
+        # Regression test for gh-4341
+        # Unpacking a scalar should give zero-dim output,
+        # even if dtype is structured
+        txt = TextIO("1")
+        dt = {'names': ('a',), 'formats': ('<i4',)}
+        x = np.genfromtxt(txt, dtype=dt, unpack=True)
```

Comment:
Here's what I get with this example (using `StringIO` instead of the internal `TextIO`):
```
In [37]: dt = np.dtype({'names': ('a',), 'formats': ('<i4',)})

In [38]: x = np.genfromtxt(StringIO('1'), dtype=dt, unpack=True)

In [39]: x                                                                                                        
Out[39]: array((1,), dtype=[('a', '<i4')])
```
I don't think the `dtype` of the return value should still be the structured dtype in this case--`unpack=True` is supposed to unpack the fields.  So I expect this output to be `array(1, dtype=int32)`.  In other words, the structure should still be pulled apart even when `len(names) == 1`.

Reply:
Good point. I modified the logic to return the single ndarray when `len(names) == 1` and a list of ndarrays when `len(names > 1)`, which I think is a little bit cleaner. Technically, `len(names) == 0)` is still possible with an empty input file and a structured dtype, but this already gives a UserWarning.